File: dtrng-javascript/Day11.js
const ops = {
  n: ({x, y}) => ({x, y: y + 1}),
  ne: ({x, y}) => ({x: x + 1, y: y + (x % 2 === 0 ? 0 : 1)}),
  se: ({x, y}) => ({x: x + 1, y: y + (x % 2 === 0 ? -1 : 0)}),
  s: ({x, y}) => ({x, y: y - 1}),
  sw: ({x, y}) => ({x: x - 1, y: y + (x % 2 === 0 ? -1 : 0)}),
  nw: ({x, y}) => ({x: x - 1, y: y + (x % 2 === 0 ? 0 : 1)}),
}

const solveA = ({data}) => {
  const {x, y} = data
    .split(',')
    .reduce(({x, y}, dir) => {
      return ops[dir]({x, y})
    }, {x: 0, y: 0})
  return Math.abs(x) + Math.max(Math.abs(y) - Math.floor((Math.abs(x) + (y >= 0 ? 0 : x % 2 === 0 ? 0 : 1)) / 2), 0)
}

const solveB = ({data}) => {
  return data
    .split(',')
    .reduce((acc, dir) => {
      const {x, y} = ops[dir](acc)
      const max = Math.abs(x) + Math.max(Math.abs(y) - Math.floor((Math.abs(x) + (y >= 0 ? 0 : x % 2 === 0 ? 0 : 1)) / 2), 0)
      return {x, y, max: max > acc.max ? max : acc.max}
    }, {x: 0, y: 0, max: 0})
    .max
}

require('fs').readFile('input.txt', 'utf8', (err, data) => {
  if (err) { console.error(err); return }
  data = data.split('\n')[0]
  console.log('solution to a:', solveA({data}))
  console.log('solution to b:', solveB({data}))
})

File: digidis-go/day11/day11.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func main() {
	row, _ := ioutil.ReadFile("input.txt")
	var dist, maxdist int
	var q, r int
	for _, s := range strings.Split(string(row), ",") {
		switch s {
		case "n":
			q--
		case "s":
			q++
		case "ne":
			r++
			q--
		case "sw":
			r--
			q++
		case "nw":
			r--
		case "se":
			r++
		}

		dist = (abs(q) + abs(r) + abs(-q-r)) / 2

		if dist > maxdist {
			maxdist = dist
		}
	}

	fmt.Printf("Distance %v\n", dist)
	fmt.Printf("Max distance %v\n", maxdist)
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
